### DISCUSSION TITLE
Upgrading Yarn to V2 and Dropping Lerna

### DISCUSSION THREAD

**Discussion by Spacesai1or** [ethereum/web3.js#4464] [General]
- `1.x`
    uses
        - `npm`
        - `lerna` (in [fixed](https://github.com/lerna/lerna#fixedlocked-mode-default) mode)
- `4.x`
    uses
        - `yarn`: package manager, with workspace support (handles local package linking and not duplicating package dependency  installs)
        - `lerna` (in [independent](https://github.com/lerna/lerna#independent-mode) mode): handles running `package.json` scripts for each package (can be replaced by yarn workspaces), but does not handle package versioning and publishing (currently it's a [manual process](https://github.com/ChainSafe/web3.js/blob/4.x-old/RELEASE.md))

---

These are the decisions that need to be made for `4.x`:

1. Are we upgrading to Yarn v2?
    - Judging solely from [this](https://shift.infinite.red/yarn-1-vs-yarn-2-vs-npm-a69ccf0229cd) article, it would seem wiser to stick with v1
2. Are we dropping the use of Lerna?
    - Regarding running `package.json` scripts for each package, Yarn workspaces provides this feature
    - If we decide to revert the decision of using the `independent` method (and use `fixed` as in `1.x`), we get the benefit of Lerna managing updating package versions and publishing each package to NPM
        - Keep in mind that the decision to move away from this was because a single breaking change in one package means all version numbers for every package must have a major version bump (even if they aren't affected by the breaking change)
    - If we stay with `independent`, it is my naive recommendation to drop the use of Lerna and just write simple scripts to update package versions and publish to NPM (although I have no idea what kind of can of worms may be awaiting us)
        - Talking with @mpetrunic, Lodestar used the `independent` method and they found it to be a headache to deal with updating package versions for each dependent package, and decided to revert to `fixed`

**Comment by mpetrunic** [comment]
> Judging solely from this article, it would seem wiser to stick with v1

That pretty outdated article, it was testing first release candidate of yarn2 :)
https://spin.atomicobject.com/2021/02/17/switch-yarn-2/
https://www.sitepoint.com/yarn-vs-npm/
